server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -941,44 +914,22 @@ private static Metadata metadataForSnapshot(SnapshotsInProgress.Entry snapshot,
         String repository,
         List<String> snapshots
     ) {
-        if (snapshotsInProgress == null || snapshotsInProgress.entries().isEmpty()) {
+        if (snapshotsInProgress == null || snapshotsInProgress.isEmpty()) {
             return Collections.emptyList();
         }
         if ("_all".equals(repository)) {
-            return snapshotsInProgress.entries();
-        }
-        if (snapshotsInProgress.entries().size() == 1) {
-            // Most likely scenario - one snapshot is currently running
-            // Check this snapshot against the query
-            SnapshotsInProgress.Entry entry = snapshotsInProgress.entries().get(0);
-            if (entry.snapshot().getRepository().equals(repository) == false) {
-                return Collections.emptyList();
-            }
-            if (snapshots.isEmpty() == false) {
-                for (String snapshot : snapshots) {
-                    if (entry.snapshot().getSnapshotId().getName().equals(snapshot)) {
-                        return snapshotsInProgress.entries();
-                    }
-                }
-                return Collections.emptyList();
-            } else {
-                return snapshotsInProgress.entries();
-            }
+            return snapshotsInProgress.asStream().collect(Collectors.toUnmodifiableList());
+        }
+        if (snapshots.isEmpty()) {
+            return snapshotsInProgress.forRepo(repository);
         }
         List<SnapshotsInProgress.Entry> builder = new ArrayList<>();
-        for (SnapshotsInProgress.Entry entry : snapshotsInProgress.entries()) {
-            if (entry.snapshot().getRepository().equals(repository) == false) {
-                continue;
-            }
-            if (snapshots.isEmpty() == false) {
-                for (String snapshot : snapshots) {
-                    if (entry.snapshot().getSnapshotId().getName().equals(snapshot)) {
-                        builder.add(entry);
-                        break;
-                    }
+        for (SnapshotsInProgress.Entry entry : snapshotsInProgress.forRepo(repository)) {
+            for (String snapshot : snapshots) {
+                if (entry.snapshot().getSnapshotId().getName().equals(snapshot)) {
+                    builder.add(entry);
+                    break;
```

Comment:
nit: now we can just do this, no need for the intermediate builder and wrapper:

```suggestion
                    return List.of(entry);
```

Reply:
I don't think we can, we have two nested loops here and we're only breaking the inner one.